specs/core/0_beacon-chain.md

```diff
@@ -1656,14 +1659,14 @@ If `state.slot % ETH1_DATA_VOTING_PERIOD == 0`:
 
 * Set `state.previous_justified_slot = state.justified_slot`.
 * Set `state.justification_bitfield = (state.justification_bitfield * 2) % 2**64`.
-* Set `state.justification_bitfield |= 2` and `state.justified_slot = state.slot - 2 * EPOCH_LENGTH` if `3 * previous_epoch_boundary_attesting_balance >= 2 * total_balance`.
-* Set `state.justification_bitfield |= 1` and `state.justified_slot = state.slot - 1 * EPOCH_LENGTH` if `3 * current_epoch_boundary_attesting_balance >= 2 * total_balance`.
+* Set `state.justification_bitfield |= 2` and `state.justified_slot = previous_epoch_start_slot` if `3 * previous_epoch_boundary_attesting_balance >= 2 * total_balance`.
+* Set `state.justification_bitfield |= 1` and `state.justified_slot = current_epoch_start_slot` if `3 * current_epoch_boundary_attesting_balance >= 2 * total_balance`.
 
 Set `state.finalized_slot = state.previous_justified_slot` if any of the following are true:
 
-* `state.previous_justified_slot == state.slot - 2 * EPOCH_LENGTH and state.justification_bitfield % 4 == 3`
-* `state.previous_justified_slot == state.slot - 3 * EPOCH_LENGTH and state.justification_bitfield % 8 == 7`
-* `state.previous_justified_slot == state.slot - 4 * EPOCH_LENGTH and state.justification_bitfield % 16 in (15, 14)`
+* `state.previous_justified_slot == previous_epoch_start_slot and state.justification_bitfield % 4 == 3`
+* `state.previous_justified_slot == previous_epoch_start_slot - 1 * EPOCH_LENGTH and state.justification_bitfield % 8 == 7`
+* `state.previous_justified_slot == previous_epoch_start_slot - 2 * EPOCH_LENGTH and state.justification_bitfield % 16 in (15, 14)`
```

Comment:
Weren't we going to fix https://github.com/ethereum/eth2.0-specs/issues/410 at the same time? Or did you want to get this PR merged first and then do a separate one for that?

Reply:
Closed this PR. I think https://github.com/ethereum/eth2.0-specs/pull/480 is the way to go. I will fix #410 when we decide which of the two approaches to go with.

#480 does the epoch transition at the _start_ of slots (before block processing) at which `slot % EPOCH_LENGTH == 0`. this avoids all the nasty -1/+1 logic found in this PR.